packages/headings/src/headings.ts

```diff
@@ -66,12 +69,38 @@ export const rehypePlugin: HeadingPlugin = ({ slug = generateSlug } = {}) => (as
   vfile.data.headings = headings
 }
 
+function initCounter (ast: any) {
+  const { children } = ast as Parent
+  const counter = new Map<string, number>()
+  // add id in the counter first because when id is set it is used as slug
+  children.forEach((child: any) => {
+    const id = child.properties?.id
+    if (id) counter.set(id, 1)
+  })
+  return counter
+}
```

Comment:
Let's skip this step, as it's not recursive, so ids in nested elements won't be detected, and this check doesn't prevent that an id outside the document (in a layout) conflicts with the headings.

The overhead of walking the entire document recursively is not acceptable for this plugin (though you can create a separate rehype plugin if you need to achieve that).